Basic implementation of the StdIO extension for catching stdin/out (#227)
junit-pioneer/junit-pioneer: docs/standard-input-output.adoc

:page-title: Standard input/output
:page-description: JUnit Jupiter extension for catchin standard input or output.

The `StdIOExtension` adds a simple way to test your classes that read from the standard
input (`System.in`) or write to the standard output (`System.out`).

== How to use the extension

You annotate your test class with `@ExtendWith(StdIOExtension.class)`. This lets JUnit know
that you would like to use the special parameters provided by the extension in your tests.
Then, when writing your tests, you can pass two arguments provided by the extension:

StdIn::

> This is for catching input from the standard in. If you are using this parameter, you have to
> annotate you method with `@StdInSource` which takes a String array and provides the values that
> you would like to read instead of waiting for input on the standard input.

Reviewer: ```suggestion
This is for simulating input from the standard in. If you are using this parameter, you have to
annotate you method with `@StdInSource`. This annotation takes a String array, which holds values that
you would like to read instead of waiting for input on the standard input. Each value of the array represents one (line of) input.
```

> This would look like something like this:

Reviewer: ```suggestion
The following example shows how to pass the values `Hello` and `World` as standard input:
```

[source,java]
----
    @ExtendWith(StdIOExtension.class)
    public class StandardInputOutputTests {
        ...
    @Test
    @StdInSource({"Hello", "World"})
    void test(StdIn in) {
        ...

----

StdOut::

> This is for catching output going to the standard out. This parameter does not require any special annotations
> (except for the `@ExtendWith(StdIOExtension.class)` on your test class).

Reviewer: ```suggestion
This is for catching output going to the standard out. This parameter does not require any additional annotations 
except for the `@ExtendWith(StdIOExtension.class)` on your test class (see below).
```

> This would look like something like this:

Reviewer: ```suggestion
```

[source,java]
----
    @ExtendWith(StdIOExtension.class)
    public class StandardInputOutputTests {
        ...
    @Test
    void test(StdOut out) {
        ...

----

> After calling your method that reads or writes to the standard in or out, you can call
> `capturedLines()` (available in both `StdIn` and `StdOut`) to verify the interactions.
> This contains the redirected in- or output as lines.

Reviewer: ```suggestion
To check your method under tests interactions with `StdIn` or `StdOut`, you can call the 
`capturedLines()` method (available in both `StdIn` and `StdOut`) inside your test and to verify the lines read by `StdIn` or written by `StdOut`.
This array contains the redirected in- or output as lines.
```

For example, after calling `System.out.print("Hello")` and `System.out.println("World")`,

> the `capturedLines()` method would contain "HelloWorld".

Reviewer: ```suggestion
the `capturedLines()` method would provides you an array containing the two strings "Hello" and "World".
```

Author: This is incorrect, it contains a single String ("HelloWorld") because you only printed one line break to the standard out.

Reviewer: Oh you right I read _System.out.print**ln**_ both times - my mistake!

:page-title: Standard input/output
:page-description: JUnit Jupiter extension for simulating standard input or catching standard output.

The `StdIOExtension` adds a simple way to test your classes that read from the standard
input (`System.in`) or write to the standard output (`System.out`).

== How to use the extension

> If you would like to use the special parameters provided by the extension

Reviewer 2: i just want to raise a discussion here, about our documentation style, or more about unifying it - and maybe find a guideline on how to write docs. Currently it seems like we have two different approaches:

1. the personal/informal one, in which we address the developer and his work directly with you and your tests (like here)
2. the formal one, in which it seems like we try to avoid such constructs (eg. RangeSource)

For a good comprehensive documentation i think we should agree on one style to use. As it makes reading the docs easier to grasp and shows some kind of consistency. The questions are

- is this something which only bothers me? than lets continue, as i did not say anything :)
- what would be the general preference. I tend to use the formal one, simply because it feels more 'technical' to me.

What is the general take on this?

Reviewer: Good spot, but I won't discuss that here, as we most probably have to align multiple places. So I'll create an bishue for this.

> If you would like to use the special parameters provided by the extension

Reviewer 2: i think we do favor one sentence per line in our documentation, as it is easier to maintain and to move around (no line length). @nicolaiparlog explained the reason in the last pioneer stream, i think he know all the words, and has the best words, so hopefully he can explain that better in written words.

so i think we should stick to one sentence per line, without breaking.

Reviewer: Yes I agree. The ~70 character line with manual wrap is only for commit messages, but not for .adoc-files.

just annotate your test class with `@ExtendWith(StdIOExtension.class)`.
Then, when writing your tests, you can pass the following two classes as arguments (provided by the extension):

StdIn(`org.junitpioneer.jupiter.StdIOExtension.StdIn`)::
This is for simulating input from the standard in. If you are using this parameter, you have to

> annotate you method with `@StdInSource`. This annotation takes a String array which holds the values that

Reviewer 2: ```suggestion
annotate your method with `@StdInSource`. This annotation takes a String array which holds the values that
```

you would like to read instead of waiting for input on the standard input. Each value of the array
represents one line of input.

The following example shows how to pass the values `Hello` and `World` as standard input:

[source,java]
----
    @ExtendWith(StdIOExtension.class)
    public class StandardInputOutputTests {
        ...
    @Test
    @StdInSource({"Hello", "World"})
    void test(StdIn in) {
        ...

----

StdOut(`org.junitpioneer.jupiter.StdIOExtension.StdOut`)::
This is for catching output going to the standard out. This parameter does not require any additional annotations
except for the `@ExtendWith(StdIOExtension.class)` on your test class (see below).

[source,java]
----
    @ExtendWith(StdIOExtension.class)
    public class StandardInputOutputTests {
        ...
    @Test
    void test(StdOut out) {
        ...

----

To check your method under tests interactions with `StdIn` or `StdOut`, you can call
`capturedLines()` (available in both `StdIn` and `StdOut`) inside your test to verify
the lines read by `StdIn` or written by `StdOut`.
This array contains the redirected in- or output as lines.

> For example, after calling `System.out.print("Hello")` and `System.out.println("World")`,
> the `capturedLines()` method would contain an array of the String "HelloWorld" because
> the first method does not print a line break.

Reviewer: ```suggestion
For example, after calling `System.out.print("Hello")` and `System.out.println("World")`,
the `capturedLines()` method would contain an array of the String "HelloWorld", because
the first method does not print a line break.
```
